x-pack/legacy/plugins/infra/server/lib/alerting/metric_threshold/register_metric_threshold_alert_type.ts

```diff
@@ -0,0 +1,143 @@
+import { schema } from '@kbn/config-schema';
+import { Server } from 'hapi';
+import {
+  MetricThresholdAlertTypeParams,
+  Comparator,
+  AlertStates,
+  METRIC_THRESHOLD_ALERT_TYPE_ID,
+} from './types';
+import { AlertServices } from '../../../../../alerting/server/types';
+
+const FIRED_ACTIONS = {
+  id: 'metrics.threshold.fired',
+  name: i18n.translate('xpack.metrics.alerting.threshold.fired', {
+    defaultMessage: 'Fired',
+  }),
+};
+
+async function getMetric(
+  { callCluster }: AlertServices,
+  { metric, searchField, aggregation, interval, indexPattern }: MetricThresholdAlertTypeParams
+) {
+  const searchBody = {
+    query: {
+      bool: {
+        filter: [
+          {
+            range: {
+              '@timestamp': {
+                gte: `now-${interval}`,
+              },
+            },
+          },
+          {
+            bool: {
+              should: [
+                {
+                  match_phrase: {
+                    [searchField.name]: searchField.value,
+                  },
+                },
+                {
+                  exists: {
+                    field: metric,
+                  },
+                },
+              ],
+            },
+          },
+        ],
+      },
+    },
+    size: 0,
+    aggs: {
+      aggregatedIntervals: {
+        date_histogram: {
+          field: '@timestamp',
+          calendar_interval: interval,
+        },
+        aggregations: {
+          aggregatedValue: {
+            [aggregation]: {
+              field: metric,
+            },
+          },
+        },
+      },
+    },
+  };
+
+  const result = await callCluster('search', {
+    body: searchBody,
+    index: indexPattern,
+  });
+
+  const { buckets } = result.aggregations.aggregatedIntervals;
+  const { value } = buckets[buckets.length - 1].aggregatedValue;
+  return value;
+}
+
+const comparatorMap = {
+  [Comparator.GT]: (a: number, b: number) => a > b,
+  [Comparator.LT]: (a: number, b: number) => a < b,
+  [Comparator.GT_OR_EQ]: (a: number, b: number) => a >= b,
+  [Comparator.LT_OR_EQ]: (a: number, b: number) => a <= b,
+};
+
+export async function registerMetricThresholdAlertType(server: Server) {
+  const { alerting } = server.plugins;
+  if (!alerting) {
+    throw new Error(
+      'Cannot register metric threshold alert type.  Both the actions and alerting plugins need to be enabled.'
+    );
+  }
+
+  alerting.setup.registerType({
+    id: METRIC_THRESHOLD_ALERT_TYPE_ID,
+    name: 'Metric Alert - Threshold',
+    validate: {
+      params: schema.object({
+        threshold: schema.number(),
+        comparator: schema.string(),
+        aggregation: schema.string(),
+        searchField: schema.object({
+          name: schema.string(),
+          value: schema.string(),
+        }),
+        metric: schema.string(),
+        interval: schema.string(),
+        indexPattern: schema.string(),
+      }),
+    },
+    actionGroups: [FIRED_ACTIONS],
```

Comment:
The `actionGroups` parameter is currently defined as a `string[]`, so I'm surprised this type checks, since `FIRED_ACTIONS` is an object and not a string.  Perhaps I'm missing something here tho.

Reply:
This failed to typecheck before because, [according to the docs](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/alerting/README.md#methods), `actionGroups` is typed as `Array<{id:string, name:string}>`

`scheduleActions` is still expecting a `string` though which is why this one's failing:
https://github.com/elastic/kibana/pull/57606/checks?check_run_id=446684113

Need to fix that. I guess `scheduleActions` is expecting `FIRED_ACTIONS.id`? Might want to clear up this ambiguity

Reply:
woops, yeah, sorry!  We did change the types on `actionGroups` last week or so, I forgot about that.  `scheduleActions()` should just take the `id` value, and not the `{id, name}` structure.   The `{id, name}` structure lets us associate an i18n "name" for a particular id, but that name is not used during the execution of alerts, only the `id`.

I've opened up issue https://github.com/elastic/kibana/issues/57853 to clean up the README.  Please feel free to add more to it!